Assets/MixedRealityToolkit.Providers/WindowsMixedReality/XRSDK/Controllers/WindowsMixedRealityXRSDKArticulatedHand.cs

```diff
@@ -0,0 +1,325 @@
+        {
+            handMeshObserver = await sourceState.Source.TryCreateHandMeshObserverAsync();
+        }
+#endif // WINDOWS_UWP
+
+        #region Update data functions
+
+        /// <inheritdoc />
+        public override void UpdateController(InputDevice inputDevice)
+        {
+            if (!Enabled) { return; }
+
+            base.UpdateController(inputDevice);
+
+            UpdateHandData(inputDevice);
+
+            for (int i = 0; i < Interactions?.Length; i++)
+            {
+                switch (Interactions[i].InputType)
+                {
+                    case DeviceInputType.IndexFinger:
+                        UpdateIndexFingerData(Interactions[i]);
+                        break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Update the hand data from the device.
+        /// </summary>
+        /// <param name="interactionSourceState">The InteractionSourceState retrieved from the platform.</param>
+        private void UpdateHandData(InputDevice inputDevice)
+        {
+            Hand hand;
+            if (inputDevice.TryGetFeatureValue(CommonUsages.handData, out hand))
+            {
+                foreach (HandFinger finger in handFingers)
+                {
+                    if (hand.TryGetFingerBones(finger, fingerBones))
+                    {
+                        for (int i = 0; i < fingerBones.Count; i++)
+                        {
+                            TrackedHandJoint trackedHandJoint = ConvertToTrackedHandJoint(finger, i);
+                            Bone bone = fingerBones[i];
+
+                            Vector3 position = Vector3.zero;
+                            Quaternion rotation = Quaternion.identity;
+
+                            if (bone.TryGetPosition(out position) || bone.TryGetRotation(out rotation))
+                            {
+                                // We want input sources to follow the Playspace, so fold in the playspace transform here to
+                                // put the controller pose into world space.
+                                position = MixedRealityPlayspace.TransformPoint(position);
+                                rotation = MixedRealityPlayspace.Rotation * rotation;
+
+                                unityJointPoses[trackedHandJoint] = new MixedRealityPose(position, rotation);
+                            }
+                        }
+
+                        // Unity doesn't provide a palm joint, so we synthesize one here
+                        unityJointPoses[TrackedHandJoint.Palm] = CurrentControllerPose;
+                    }
+                }
+
+                CoreServices.InputSystem?.RaiseHandJointsUpdated(InputSource, ControllerHandedness, unityJointPoses);
+            }
+
+#if WINDOWS_UWP
+            if (!CoreServices.InputSystem.InputSystemProfile.HandTrackingProfile.EnableHandMeshVisualization)
```

Comment:
should we be null checking anything here?

Reply:
updated